circuits/cpp/barretenberg/cpp/src/barretenberg/proof_system/relations/ecc_vm/ecc_wnaf_relation.cpp

```diff
@@ -0,0 +1,221 @@
+    row_slice += row_slice;
+    row_slice += row_slice;
+    row_slice += w3;
+    auto sum_delta = scalar_sum * FF(1ULL << 16) + row_slice;
+    const auto check_sum = scalar_sum_new - sum_delta;
+    std::get<8>(accumulator) += precompute_select * check_sum * scaled_transition_is_zero;
+
+    /**
+     * @brief Round transition logic.
+     * Goal: `round` is an integer in [0, ... 7] that tracks how many slices we have processed for a given scalar.
+     * i.e. number of 4-bit WNAF slices processed = round * 4.
+     * We apply the following constraints:
+     * If q_transition = 0, round increments by 1 between rows.
+     * If q_transition = 1, round value at current row = 7
+     * If q_transition = 1, round value at next row = 0
+     * Question: is this sufficient? We don't actually range constrain `round` (expensive if we don't need to!).
+     * Let us analyse...
+     * 1. When `q_transition = 1`, we use a set membership check to map the tuple of (pc, scalar_sum) into a set.
+     * We compare this set with an equivalent set generated from the transcript columns. The sets must match.
+     * 2. Only case where, at row `i`, a Prover can set `round` to value > 7 is if `q_transition = 0` for all j > i.
+     *  `precompute_pc` decrements by 1 when `q_transition` = 1
+     * We can infer from 1, 2, that if `round > 7`, the resulting wnafs will map into a set at a value of `pc` that is
+     * greater than all valid msm pc values (assuming the set equivalence check on the scalar sums is satisfied).
+     * The resulting msm output of such a computation cannot be mapped into the set of msm outputs in
+     * the transcript columns (see relations in ecc_msm_relation.cpp).
+     * Conclusion: not applying a strict range-check on `round` does not affect soundness (TODO validate this!)
+     */
+    // We combine checks 0, 1 into a single relation
+    // q_transition * (round - 7) + (-q_transition + 1) * (round_shift - round - 1)
+    // => q_transition * (round - 7 - round_shift + round + 1) + (round_shift - round - 1)
+    // => q_transition * (2 * round - round_shift - 6) + (round_shift - round - 1)
+    const auto round_check = round_shift - round - 1;
+    std::get<9>(accumulator) += precompute_select * scaled_transition * ((round - round_check - 7) + round_check);
+    std::get<10>(accumulator) += precompute_select * scaled_transition * round_shift;
+
+    /**
+     * @brief Scalar transition checks.
+     * 1: if q_transition = 1, scalar_sum_new = 0
+     * 2: if q_transition = 0, pc at next row = pc at current row
+     * 3: if q_transition = 1, pc at next row = pc at current row - 1 (decrements by 1)
+     * (we combine 2 and 3 into a single relation)
+     */
+    std::get<11>(accumulator) += precompute_select * scalar_sum_new * scaled_transition;
+    // (2, 3 combined): q_transition * (pc - pc_shift - 1) + (-q_transition + 1) * (pc_shift - pc)
+    // => q_transition * (2 * (pc - pc_shift) - 1) + (pc_shift - pc)
```

Comment:
This formula doesn't match the one you have in the code.

Reply:
I...think it does? I rewrote the comment formula to better match the code. Can you double check this as I am fairly confident the formula matches

Reply:
In the code you are multiplying the whole ((2 * (pc - pc_shift) - 1) + (pc_shift - pc)) by q_transition. In the comments just the (-2* pc_delta -1)

Reply:
pc_delta = pc_shift - pc

The comment says: q_transition * (-2 * (pc_shift - pc) - 1) 
(-2 * (pc - pc_shift )- 1) = -2 * pc_delta - 1 , which is what the code does

Reply:
oh got it. Thanks for spotting that Kesha, will update.